README.md
# Gerador de container Postgis

Esse script foi criado para gerar um container Docker com o [Postgis](https://postgis.net) 
usando a imagem oficial. O difierencial é que já gera um arquivo pg_pass, evitando
 a necessidade de solicitar senha sempre que logar com o super usuário.  
E cria uma pasta no diretório corrente para receber os dados do banco.

As opções padrões do script são:

- Nome do container = meus-dados-geográficos;  
- Nome do banco = meupostgis;  
- Porta do host = 5433;  
- Root = postgres;  
- Senha do root = postgres;  
- Recria a pasta data: não;  
- Remove o container caso exista: não;  
- Pasta data será criada no diretório atual desse script;  
- Pasta conf = pasta presente nesse repositório;  
- Pasta scripts = pasta presente nesse repositório;  

Esses padrões podem ser configurados conforme as opções que estarão no help

## Modo de uso

```shel
./create_postgis.sh
```

É necessário rodar o script como super user para fazer algumas modificações.

## Help

As seguinte opções podem ser passadas:

- `--clear_data` > Remove a pasta data existente no diretório atual e cria uma nova.
- `--rm_container` > Remove o container de mesmo nome caso já exista.
- `--no_pgpass` > Não cria o arquivo pgpass

> Essas opções anteriores não precisam de parâmetros.

- `-c | --container string` > Nome do container a ser criado.  
- `-C | --config-dir path` > Caminho para a pasta conf a ser mapeada para o contaner.  
- `-d | --database string` > Nome do banco a ser criado dentro do container.  
- `-D | --data-dir path` > Caminho para o diretório onde a pasta data será criada (caso não exista) e mapeada para o contaner.  
- `-g | --gis-version string` > Tag da versão do container do postgis a ser usada.  
- `-p | --port number` > Número da porta do host que irá receber a interna do banco.  
- `-P | --password string` > Senha do usuário Root.  
- `-s | --script-dir path` > Caminho da pasta que contém o script a ser executado no banco.  
- `-U | --user string` > Nome do usuário root.  

## Pasta scripts

Esta pasta é mapeada para dentro do container (`/script`), caso queira usar algum arquivo dentro do mesmo, basta jogar nessa pasta.  
Não é necessário reiniciar o container após inserir arquivos nessa pasta.

### 10_postgis.sh

É usado para criar o banco e as extensões. caso queira criar uma extensão a mais, basta alterar esse arquivo.
Ele é mapeado para dentro do entrypoint do container para ser executado.

## Executando o arquivo de qualquer diretório

Basta adicionar um link na pasta usr/bin que estará disponível para rodar do terminal.


CHANGELOG.md
# CHANGELOG

>The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),  
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

* Tradução
* Help considerando a linguagem do sistema
* Criar um manual para ser usado no `man`

## [0.1.2] 2022-11-28

### Changed

* Versão do container docker do postgis para 15-3.3  
* Porta padrão a ser utilizada pelo container no host para 5432  

## [0.1.1] 2022-05-23

### Added

### Changed

### Fixed

* Bug de testar o acesso à pasta data antes da criação (quando a pasta não existe).

## [0.1.0] - 2022-05-20

### Added

* Opção para indicar o caminho da pasta conf com o script de configuração do banco a ser mapeada para o container.
* Opção para indicar o caminho da pasta com o script de configuração do banco.
* Opção para indicar o caminho da pasta script a ser mapeada para o banco.
* Opção para indicar tag da versão do [Postgis](https://postgis.net) a ser usada.
* Função para exibir a versão atual do script.
* Opção de exibir a versão atual do script.
* Explicação de como colocar o script no path para ser executado em qualquer lugar pelo terminal.
* Criação do arquivo de changelog.

### Changed

* Melhorias no Readme.
* Melhorias na função usage (help).
* Opção de indicar a pasta com o script de configuração do banco.
* Modificação para utilizar o diretório atual do script para utilizar as pastas padrões caso não o usuário não indique o caminho das pastas que deseja usar.

### Fixed

## [0.0.0] - 2021-07-21

Criação do script e repositório.


scripts/10_postgis.sh
#!/bin/sh

# Adaptei a idéia do murder - 
# https://gist.github.com/murder/bf5c59291e4e9fd5e06adc9ce71340fc

set -e

# Perform all actions as $POSTGRES_USER
export PGUSER="$POSTGRES_USER"

# Create the 'template_postgis' template db
"${psql[@]}" <<- 'EOSQL'
CREATE DATABASE template_postgis IS_TEMPLATE true;
EOSQL

# Load PostGIS into both template_database and $POSTGRES_DB
for DB in template_postgis "$POSTGRES_DB"; do
        echo "Loading PostGIS extensions into $DB"
        "${psql[@]}" --dbname="$DB" <<-'EOSQL'
                CREATE EXTENSION IF NOT EXISTS postgis;
                CREATE EXTENSION IF NOT EXISTS postgis_topology;
                CREATE EXTENSION IF NOT EXISTS fuzzystrmatch;
                CREATE EXTENSION IF NOT EXISTS postgis_tiger_geocoder;
                CREATE EXTENSION IF NOT EXISTS pgcrypto;

EOSQL
done


create_postgis.sh
#!/bin/bash

# Author: Kyle Felipe
# E-mail: kylefelipe at gmail.com
# data: 18/07/2021
# Ùltima atualização: 28/11/2022
# Script feito para criar um container com postgis e com a pasta data em um
# local específico, a princípio é uma pasta data no diretório atual
# É intenção futura poder escolher via opções onde colocar a pasta data.

REPOLINK="https://github.com/kylefelipe/postgis_container_generator"
version='0.1.2'
database="meupostgis"
hostname="localhost"
container_name="meus-dados-geograficos"
host_port="5432"
pg_password="postgres"
pg_user="postgres"
remove_data="n"
remove_container="n"
pg_pass="s"
data_dir="$(pwd -P)"
script_dir="$(pwd -P)/scripts"
conf_dir="$(pwd -P)/conf"
postgis_version="15-3.3"

usage() {
    echo "Uso:  sudo create_postgis.sh [OPÇÃO]

    Essas opções possuem argumentos obrigatórios:

        [ -c | --container container name ]
        [ -C | --config-dir path path to config folder to map to container ]
        [ -d | --database database name ]
        [ -D | --data-dir path to folder data to map to container ]
        [ -g | --gis-version string postgis container version tag to use, link in the end of this help ]
        [ -h | --hostname hostname hostname/ip to expose at host ]
        [ -p | --port database port ]
        [ -P | --password database password ]
        [ -s | --script-dir path path to script folder to map to container ]
        [ -U | --user database user (root) ]
    
    Essas opções não precisam de argumentos:

        [ --clear_data erases data folder ]
        [ --no_pgpass skip pgpass config]
        [ --rm_container remove container before create ]
        [ --help exibe esse help ]
        [ --version informa a versão e sai ]"
    echo ""
    echo "Cheque as tags que podem ser utilizadas no postgis em <https://hub.docker.com/r/postgis/postgis/tags>"
    echo "Página do repositório desse script: <$REPOLINK>"
    echo "Envie os erros e sugestões para <$REPOLINK/issues>"
    echo "Se foi útil, deixe uma estrelinha"
    echo "LLP _\\\\//"
    echo "<www.kylefelipe.com>"
    exit 2
}

use_pgpass() {
    echo "Criando aquivo .pgpas na pasta $conf_dir"
    echo "localhost:5432:$database:$pg_user:$pg_password" > $conf_dir/.pgpass
    chmod 0600 $conf_dir/.pgpass
}

version() {
    echo $version
    exit 2
}

PARSED_ARGUMENTS=$(getopt -a -n argument -o h:c:C:d:D:f:p:P:U:v: \
                    --long container_name:,config-dir:,database:,data-dir:,hostname:,port:,password:,script-dir:,user:,clear_data,rm_container,help,no_pgpass,version -- "$@")

VALID_ARGUMENTS=$?

if [ "$VALID_ARGUMENTS" != "0" ]; then
    usage
fi

eval set -- "$PARSED_ARGUMENTS"

while :
do
    case "$1" in
    -c | --container)
        container_name="$2"
        shift 2
    ;;
    -C | --config-dir)
        config_path="$2"
        shift 2
    ;;
    -d | --database)
        database="$2"
        shift 2
    ;;
    -D | --data-dir)
        data_dir="$2"
        shift 2
    ;;
    -g | --gis-version)
        postgis_version="$2"
        shift 2
    ;;
    -h | --hostname)
        hostname="$2"
        shift 2
    ;;
    -p | --port)
        host_port="$2"
        shift 2
    ;;
    -P | --password)
        pg_password="$2"
        shift 2
    ;;
    -s | --script-dir)
        script_dir="$2"
        shift 2
    ;;
    -U | --user)
        pg_user="$2"
        shift 2
    ;;
    --clear_data)
        remove_data="s"
        shift
    ;;
    --rm_container)
        remove_container="s"
        shift
    ;;
    --no_pgpas)
        pg_pass="n"
        shift
    ;;
    --help)
        usage
        shift 2
    ;;
    -v | --version)
        version
        shift 2
    ;;
    --)
        shift
        break
        ;;
    *)
        echo "Opção $1 não reconhecida."
        usage
        ;;
    esac
done

if [ "$remove_data" = "s" ] && [ -d "$data_dir/data" ]; then
    echo "Removendo a pasta data."
    rm -rf "$data_dir/data"
else
    echo "Pasta $data_dir/data inexistente!"
fi

if [ "$remove_container" = "s" ]; then
    echo "Removendo container $container_name"
    docker container rm -f "$container_name"
fi

if [ ! -d "$data_dir/data" ]; then
    echo "Criando a pasta /data dentro do diretório $data_dir"
    mkdir -p "$data_dir/data"
    echo "Pronto!"

fi

if [ -d "$data_dir/data" ] && [ ! -w "$data_dir/data" ] && [ ! -x "$data_dir/data" ]; then
    echo "Usuário não tem permissão para alterar a pasta $data_dir/data"
    echo "Considere executar como super usuário!"
    exit 1
fi


if [ "$VALID_ARGUMENTS" = "0" ]
then
    existing_container="$(docker ps -q -f name=$container_name)"
    if [ -n "$existing_container" ] && [ "$remove_container" = "n" ]; then
        echo "Já existe um container com o nome $container_name."
        echo "Por favor, especifique um novo nome de container ou remova o já exstente"
        echo "ou use a opção --rm_container, para remover um container pré existente de mesmo nome"
        usage
    fi

    echo ""
    if [ "$pg_pass" = "s" ]; then
        use_pgpass
    fi

    echo ""
    echo "Criando o container $container_name em modo daemon."
    echo ""
    echo "Imagem Postgis utilizada: postgis/postgis:$postgis_version"
    echo "https://hub.docker.com/r/postgis/postgis"

    sudo docker run -d \
        --restart unless-stopped --name "$container_name" \
        -p 0.0.0.0:"$host_port":5432 \
        -e POSTGRES_PASSWORD="$pg_password" \
        -e POSTGRES_USER="$pg_user" \
        -e POSTGRES_DB="$database" \
        -e PGDATA=/var/lib/postgresql/data/pgdata \
        --ipc=host \
        -v "$data_dir/data":/data \
        -v "$script_dir":/scripts \
        -v "$script_dir"/10_postgis.sh:/docker-entrypoint-initdb.d/10_postgis.sh \
        -v "$conf_dir"/.pgpass:/root/.pgpass \
        -v "$data":/var/lib/postgresql/data/pgdata \
        postgis/postgis:"$postgis_version"
    
    echo ""
    if [ "$(docker ps -q -f name=$container_name -f status=running)"  == "" ]; then
        echo -n "Aguardando container iniciar"
        while [ "$(docker ps -q -f name=$container_name -f status=running)"  == "" ]; do
            echo -n "."
            sleep 1
        done
    fi

    sleep 10

    echo ""
    echo "PostgreSQL Version:"
    docker exec "$container_name" pg_config --version

    echo ""
    echo "Postgis Version:"
    docker container exec "$container_name" psql -U "$pg_user" -d "$database" -c 'SELECT PostGIS_Lib_Version();'

    echo ""
    echo "Container criado com sucesso!"
    echo "Para acessar basta conectar em:"
    echo "${hostname}:$host_port"
    echo ""
    echo "Para parar o container:"
    echo "docker container stop $container_name"
    echo ""
    echo "Para iniciar o container (reutilizar):"
    echo "docker container start $container_name"
    echo ""
    echo "Para acessar diretamente o psql dentro do container:"
    echo "docker container exec -it $container_name psql -U $pg_user -d $database"
    echo ""
    echo "Para acessar o shell do container:"
    echo "docker container exec -it $container_name bash"
    echo ""
    echo "Be Happy!"
    echo "LLP _\\\\//"
fi
